sources/file.go

```diff
@@ -0,0 +1,247 @@
+const InnerPathSeparator = "!"
+
+type seekReaderAt interface {
+	io.ReaderAt
+	io.Seeker
+}
+
+// File is a source for yielding fragments from a file or other reader
+type File struct {
+	// Content provides a reader to the file's content
+	Content io.Reader
+	// Path is the resolved real path of the file
+	Path string
+	// Symlink represents a symlink to the file if that's how it was discovered
+	Symlink string
+	// Buffer is used for reading the content in chunks
+	Buffer []byte
+	// Config is the gitleaks config used for shouldSkipPath. If not set, then
+	// shouldSkipPath is ignored
+	Config *config.Config
+	// outerPaths is the list of container paths (e.g. archives) that lead to
+	// this file
+	outerPaths []string
+	// MaxArchiveDepth limits how deep the sources will explore nested archives
+	MaxArchiveDepth int
+	// archiveDepth is the current archive nesting depth
+	archiveDepth int
+}
+
+// Fragments yields fragments for the this source
+func (s *File) Fragments(ctx context.Context, yield FragmentsFunc) error {
+	format, _, err := archives.Identify(ctx, s.Path, nil)
+	if err == nil && format != nil {
+		if s.archiveDepth+1 > s.MaxArchiveDepth {
+			logging.Warn().Str(
+				"path", s.FullPath(),
+			).Int(
+				"max_archive_depth", s.MaxArchiveDepth,
+			).Msg("skipping archive: exceeds max archive depth")
+			return nil
+		}
+		if extractor, ok := format.(archives.Extractor); ok {
+			return s.extractorFragments(ctx, extractor, s.Content, yield)
+		}
+		if decompressor, ok := format.(archives.Decompressor); ok {
+			return s.decompressorFragments(decompressor, s.Content, yield)
+		}
+		logging.Warn().Str("path", s.FullPath()).Msg("skipping unknown archive type")
+	}
+
+	return s.fileFragments(bufio.NewReader(s.Content), yield)
+}
+
+// extractorFragments recursively crawls archives and yields fragments
+func (s *File) extractorFragments(ctx context.Context, extractor archives.Extractor, reader io.Reader, yield FragmentsFunc) error {
+	if _, isSeekReaderAt := reader.(seekReaderAt); !isSeekReaderAt {
+		switch extractor.(type) {
+		case archives.SevenZip, archives.Zip:
+			tmpfile, err := os.CreateTemp("", "gitleaks-archive-")
+			if err != nil {
+				logging.Error().Str("path", s.FullPath()).Msg("could not create tmp file")
+				return nil
+			}
+			defer func() {
+				_ = tmpfile.Close()
+				_ = os.Remove(tmpfile.Name())
+			}()
+
+			_, err = io.Copy(tmpfile, reader)
+			if err != nil {
+				logging.Error().Str("path", s.FullPath()).Msg("could not copy archive file")
+				return nil
+			}
+
+			reader = tmpfile
+		}
+	}
+
+	return extractor.Extract(ctx, reader, func(_ context.Context, d archives.FileInfo) error {
+		if d.IsDir() {
+			return nil
+		}
+
+		innerReader, err := d.Open()
+		if err != nil {
+			logging.Error().Str("path", s.FullPath()).Err(err).Msg("could not open archive inner file")
+			return nil
+		}
+		path := filepath.Clean(d.NameInArchive)
+
+		if s.Config != nil && shouldSkipPath(s.Config, path) {
```

Comment:
`innerReader.Close()` should be a deferred statement after the err check. Presently, I believe this will cause a leak if this is true.